src/mscorlib/src/System/Text/StringBuilder.CoreCLR.cs

```diff
@@ -29,21 +27,42 @@ internal unsafe void ReplaceBufferInternal(char* newBuffer, int newLength)
             if (newLength > m_MaxCapacity)
                 throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
 
-            // Leave space for a NUL character at the end, the previous implementation did this
```

Comment:
It wasn't clear to me why this space for a null char was necessary, since as far as I can tell the only time it's added is when doing this marshaling out, which means it can't be a general requirement of StringBuilder.  So I removed it.  But let me know if I missed something.

Reply:
I believe this is called coming back here: https://github.com/dotnet/coreclr/blob/acb110d3e9218bce176b411e32a22f8cb7abb1b0/src/vm/ilmarshalers.cpp#L864

Reply:
Yes.  Why is that relevant?  What am I missing?

Reply:
I assumed it mattered if you were to pass the builder back out again after copying in the returned chars (i.e. guaranteed to be null terminated). There is a good chance I'm confusing what is going on here though.

Reply:
Back when I converted these methods from native to managed I could not find any use for the NUL character, it's most likely a leftover from early .NET days when `StringBuilder` had a very different implementation. But removing it changes the resulting builder's capacity and that's a bit risky because the capacity is checked by builder's `Equals` method. The test failure you see is AFAIR the result of that.

Reply:
> But removing it changes the resulting builder's capacity and that's a bit risky because the capacity is checked by builder's Equals method. The test failure you see is AFAIR the result of that.

Ah, interesting, thanks for pointing that out, @mikedn.

I'd be interested in others' opinions, but I still think this is a reasonable change to take, even if it means StringBuilder.Capacity ends up returning a different result than it previously would have, and thus Equals (which is documented to factor in Capacity) could end up returning a different result than it would have.  I'd argue that, even though it's visible, the capacity a collection chooses to use internally is an implementation detail that's subject to change.

My concern is that, for better or worse, there's a decent amount of code that does:
```C#
var sb = new StringBuilder(expectedSize);
PInvoke(sb, sb.Capacity);
return sb.ToString();
```
and if we then force the result to actually require expectedSize+1 chars, we end up needing to allocate another char[], whereas we wouldn't with just expectedSize.

@jkotas, @vancem, any compat opinions here?

Reply:
> PInvoke(sb, sb.Capacity);

Hmm, the interesting thing about the typical use case is that the capacity must already include the NUL terminator. Because of this we won't end up with a larger capacity by including the NUL character in the new capacity calculation.

We'll get differences only when StringBuilder is used as a return type or if space for NUL is manually added - `PInvoke(sb, sb.Capacity + 1)`.

Also, the later variant can produce a builder where `Capacity > MaxCapacity`:
```C#
var sb = new StringBuilder(83, 83);
Console.WriteLine(GetCurrentDirectoryW(sb.Capacity + 1, sb));
Console.WriteLine(sb.Length);
Console.WriteLine(sb.Capacity);
Console.WriteLine(sb.MaxCapacity);
```
On my machine this displays
```
83
83
84
83
```

Reply:
> the interesting thing about the typical use case is that the capacity must already include the NUL terminator

Yeah, I guess that's true.  The API let's say is specified to accept a buffer of size N large enough to hold N-1 chars plus a null terminator, you specify the size of the buffer as N, upon return we calculate the resulting string length as N-1, and then we add back a char for the null terminator to be back at N.

Ok, I'll put back the +1.